src/http_req.c

```diff
@@ -0,0 +1,202 @@
+            free(h->value);
+        }
+        free(req->resp.headers);
+    }
+    if (req->resp.status) {
+        free(req->resp.status);
+    }
+    free(req->path);
+    free(req->method);
+}
+
+size_t http_req_process(um_http_req_t *req, const char* buf, ssize_t len) {
+    UM_LOG(TRACE, "processing \n%.*s", len, buf);
+    size_t processed = http_parser_execute(&req->parser, &HTTP_PROC, buf, len);
+    UM_LOG(VERB, "processed %zd out of %zd", processed, len);
+    return processed;
+}
+
+static void free_hdr(um_http_hdr *hdr) {
+    free(hdr->name);
+    free(hdr->value);
+}
+
+void free_hdr_list(um_header_list *l) {
+    um_http_hdr *h;
+    while (!LIST_EMPTY(l)) {
+        h = LIST_FIRST(l);
+        LIST_REMOVE(h, _next);
+
+        free_hdr(h);
+        free(h);
+    }
+}
+
+size_t http_req_write(um_http_req_t *req, char *buf, size_t maxlen) {
+    size_t len = snprintf(buf, maxlen,
+                          "%s %s HTTP/1.1\r\n",
+                          req->method, req->path);
+
+    if (strcmp(req->method, "POST") == 0 ||
+        strcmp(req->method, "PUT") == 0 ||
+        strcmp(req->method, "PATCH") == 0) {
+        if (!req->req_chunked && req->req_body_size == -1) {
+            size_t req_len = 0;
+            struct body_chunk_s *chunk = req->req_body;
+            while (chunk != NULL) {
+                req_len += chunk->len;
+                chunk = chunk->next;
+            }
+            req->req_body_size = req_len;
+            char length_str[16];
+            sprintf(length_str, "%ld", req_len);
+            set_http_header(&req->req_headers, "Content-Length", length_str);
+        }
+    }
+
+    um_http_hdr *h;
+    LIST_FOREACH(h, &req->req_headers, _next) {
+        len += snprintf(buf + len, maxlen - len, "%s: %s\r\n", h->name, h->value);
+    }
+
+    len += snprintf(buf + len, maxlen - len, "\r\n");
+    return len;
+}
+
+void set_http_header(um_header_list *hl, const char* name, const char *value) {
+    um_http_hdr *h;
+    LIST_FOREACH(h, hl, _next) {
+        if (strcmp(h->name, name) == 0) {
+            break;
+        }
+    }
+
+    if (value == NULL) {
+        if (h != NULL) {
+            LIST_REMOVE(h, _next);
+            free(h->value);
+            free(h->name);
+        }
+        return;
+    }
+
+    if (h == NULL) {
+        h = malloc(sizeof(um_http_hdr));
+        h->name = strdup(name);
+        LIST_INSERT_HEAD(hl, h, _next);
+    } else {
+        free(h->value);
+    }
+
+    h->value = strdup(value);
+}
+
+static int http_headers_complete_cb(http_parser *p) {
+    um_http_req_t *req = p->data;
+    req->state = headers_received;
+    if (req->resp_cb != NULL) {
+        req->resp_cb(&req->resp, req->data);
+    }
+    return 0;
+}
+
+static int http_header_field_cb(http_parser *parser, const char *f, size_t len) {
+    um_http_req_t *req = parser->data;
+    if (req->resp.headers == NULL) {
+        req->resp.headers = calloc(20, sizeof(um_http_hdr));
```

Comment:
is this realloc'd somewhere if over 20 (or errored out at least)?

Reply:
changed response headers into `um_headers_list`